presto-orc/src/test/java/com/facebook/presto/orc/TestOrcReaderMemoryUsage.java

```diff
@@ -0,0 +1,329 @@
+import static com.facebook.presto.orc.OrcTester.createOrcRecordWriter;
+import static com.facebook.presto.orc.OrcTester.createSettableStructObjectInspector;
+import static com.facebook.presto.spi.block.MethodHandleUtil.compose;
+import static com.facebook.presto.spi.block.MethodHandleUtil.nativeValueGetter;
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.spi.type.VarcharType.VARCHAR;
+import static io.airlift.testing.Assertions.assertGreaterThan;
+import static io.airlift.testing.Assertions.assertGreaterThanOrEqual;
+import static org.testng.Assert.assertEquals;
+
+public class TestOrcReaderMemoryUsage
+{
+    private static final TypeManager TYPE_MANAGER = new TypeRegistry();
+
+    static {
+        // associate TYPE_MANAGER with a function registry
+        new FunctionRegistry(TYPE_MANAGER, new BlockEncodingManager(TYPE_MANAGER), new FeaturesConfig());
+    }
+
+    @Test
+    public void testVarcharTypeWithoutNulls()
+            throws Exception
+    {
+        int rows = 5000;
+        OrcRecordReader reader = null;
+        try (TempFile tempFile = createSingleColumnVarcharFile(rows, 10)) {
+            reader = createCustomOrcRecordReader(tempFile, ORC, OrcPredicate.TRUE, VARCHAR, INITIAL_BATCH_SIZE);
+            assertInitialRetainedSizes(reader, rows);
+
+            long stripeReaderRetainedSize = reader.getCurrentStripeRetainedSizeInBytes();
+            long streamReaderRetainedSize = reader.getStreamReaderRetainedSizeInBytes();
+            long readerRetainedSize = reader.getRetainedSizeInBytes();
+            long readerSystemMemoryUsage = reader.getSystemMemoryUsage();
+
+            while (true) {
+                int batchSize = reader.nextBatch();
+                if (batchSize == -1) {
+                    break;
+                }
+
+                Block block = reader.readBlock(BIGINT, 0);
+                assertEquals(block.getPositionCount(), batchSize);
+
+                // We only verify the memory usage when the batchSize reaches MAX_BATCH_SIZE as batchSize may be
+                // increasing during the test, which will cause the StreamReader buffer sizes to increase too.
+                if (batchSize < MAX_BATCH_SIZE) {
+                    continue;
+                }
+
+                // The StripeReader uses memory to hold the data, so it should increase.
+                assertGreaterThan(reader.getCurrentStripeRetainedSizeInBytes(), stripeReaderRetainedSize);
+                // We also account for the StreamReader local buffers. For SliceDictionaryStreamReader, there are two
+                // buffers: isNullVector and inDictionaryVector, and each buffer has 1024 boolean values.
+                assertGreaterThanOrEqual(reader.getStreamReaderRetainedSizeInBytes() - streamReaderRetainedSize, 2048L);
+                // The total retained size and system memory usage should be strictly larger than 2048L because of the instance sizes
```

Comment:
nit: For some comments you have a period at the end of the sentences, but not for all.